### PR TITLE
Registering enum types in data repository

### PR DESCRIPTION
**Pull request by klevzoff**
Pretty random PR, but I was looking at converting a whole bunch of solver inputs that are currently unchecked strings, into enums to improve maintainability. I was inspired by the treatment of `TimeIntegrationOption` in https://github.com/GEOSX/GEOSX/pull/885, but writing a custom conversion function for every new enum seems to not scale too well, and also leaves the question of proper schema-based validation (as opposed to just runtime checking). So the goal is to define the set of accepted values once and have everything else automated. Ideally defining a new enum type should be concise enough to not require a separate header file.

I looked at approaches of [MagicEnum](https://github.com/Neargye/magic_enum) and [Better Enums](http://aantron.github.io/better-enums/) libraries, which are both great in terms of usability. But one of them uses compiler hacks to generate strings and the other does it with a whole lot of macro substitutions (although there is only one user-facing macro), neither of which I really like.

I ended up creating a small macro that needs to be called after enum definition and associates a list of strings with the enum type. The macro works by injecting a function `getEnumStrings()` that stores the string in a local static array (so that linking issues can be avoided), as well as `operator<<` and `operator>>`, all three of which can be found by ADL since they are in the same namespace as enum. A class template `EnumStrings<T>` provides convenience functions to access the strings in various ways and perform enum <-> string conversion.

XML validation was improved by constructing an explicit validation pattern for enum inputs out of registered strings. As a result I caught quite a few outdated files under `examples` directory. A whole bunch of xml and rst files are updated, hence the size of this PR.

I took the liberty of converting some of the existing inputs that were read as strings into enums directly registered and read by the data repository machinery. Examples include `TimeIntegrationOption`, `CouplingTypeOption` in physics solvers, `SolverType` and `PreconditionerType` in linear algebra, `FluidType` in constitutive.

Rebaseline needed (only structural): https://github.com/GEOSX/integratedTests/pull/86

### CODE REVIEW

**Comment by klevzoff** [comment]
I pushed a new implementation, IMO much cleaner (and no looming linker problems). This one uses a simple macro with a list of strings to instantiate the necessary struct.

File: src/coreComponents/common/Enum.hpp [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+template<>                                                            \
+struct EnumStringsBase< ENUM >                                        \
+{                                                                     \
+  static_assert( std::is_enum< ENUM >::value, "Not an enumeration" ); \
+  static constexpr bool defined = true;                               \
+                                                                      \
+  static auto const & get()                                           \
+  {                                                                   \
+    constexpr int N = internal::countArgs( __VA_ARGS__ );             \
+    static constexpr std::array< char const *, N > ss{ __VA_ARGS__ }; \
```
Comment (klevzoff) [inline src/coreComponents/common/Enum.hpp:94]: Storing strings as local static array avoids all kinds of linking issues.

File: src/coreComponents/physicsSolvers/simplePDE/LaplaceFEM.hpp [hunk 2]
```diff
@@ -141,17 +142,15 @@ class LaplaceFEM : public SolverBase
...
-  virtual void PostProcessInput() override final;
-
 private:
 
   string m_fieldName;
   timeIntegrationOption m_timeIntegrationOption;
 
 };
 
+ENUM_STRINGS( LaplaceFEM::timeIntegrationOption, "SteadyState", "ImplicitTransient", "ExplicitTransient" )
```
Comment (klevzoff) [inline src/coreComponents/physicsSolvers/simplePDE/LaplaceFEM.hpp:152]: We can also assign strings to enums declared inside a class. The only caveat is the macro has to be called at the namespace scope.

File: src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.hpp [hunk 3]
```diff
@@ -38,6 +39,20 @@ class FieldSpecificationBase;
...
+ * The options for time integration
+ */
+enum class TimeIntegrationOption : int
+{
+  QuasiStatic,    //!< QuasiStatic
+  ImplicitDynamic,//!< ImplicitDynamic
+  ExplicitDynamic //!< ExplicitDynamic
+};
+
+ENUM_STRINGS( TimeIntegrationOption, "QuasiStatic", "ImplicitDynamic", "ExplicitDynamic" )
```
Comment (klevzoff) [inline src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.hpp:54]: Now that enum declaration is pretty short, I moved it back into the solver header. Ideally it should even be moved into the `SolidMechanicsLagrangianFEM` class scope. I may do this later.

File: src/coreComponents/common/Enum.hpp (hunk 1 above)
Comment (rrsettgast) [reply src/coreComponents/common/Enum.hpp:94]: nice!

File: examples/hydraulicFracturing/KGD/KGD.xml [hunk 4]
```diff
@@ -8,7 +8,7 @@
       name="hydrofracture"
       solidSolverName="lagsolve"
       fluidSolverName="SinglePhaseFlow"
-      couplingTypeOption="TightlyCoupled"
+      couplingTypeOption="FIM"
```
Comment (klevzoff) [inline examples/hydraulicFracturing/KGD/KGD.xml:11]: XML validation caught a whole bunch of outdated examples

File: src/coreComponents/dataRepository/ConduitRestart.hpp [hunk 5]
```diff
@@ -73,6 +73,10 @@ CONDUIT_TYPE_INFO( unsigned long long, CONDUIT_NATIVE_UNSIGNED_LONG_LONG );
 CONDUIT_TYPE_INFO( float, CONDUIT_NATIVE_FLOAT );
 CONDUIT_TYPE_INFO( double, CONDUIT_NATIVE_DOUBLE );
 
+// Enum types forward to underlying integer types
+template< typename T >
+struct conduitTypeInfo< T, std::enable_if_t< std::is_enum< T >::value > > : public conduitTypeInfo< std::underlying_type_t< T > >
+{};
```
Comment (klevzoff) [inline src/coreComponents/dataRepository/ConduitRestart.hpp:79]: This is to enable serialization of enums as their underlying type

File: src/coreComponents/fileIO/schema/schema.xsd [hunk 6]
```diff
@@ -594,19 +589,44 @@ Available options are: jacobi, blockJacobi, gaussSeidel, blockGaussSeidel, cheby
...
+* cg
+* gmres
+* fgmres
+* bicgstab
+* preconditioner-->
+		<xsd:attribute name="solverType" type="geosx_LinearSolverParameters_SolverType" default="direct" />
+	</xsd:complexType>
+	<xsd:simpleType name="geosx_LinearSolverParameters_PreconditionerType">
+		<xsd:restriction base="xsd:string">
+			<xsd:pattern value=".*[\[\]`$].*|none|jacobi|gs|sgs|iluk|ilut|icc|ict|amg|mgr|block" />
```
Comment (klevzoff) [inline src/coreComponents/fileIO/schema/schema.xsd:622]: These would be better served by `xsd:enumeration` than a regexp, but this would require more substantial changes to our schema generation code

File: src/coreComponents/physicsSolvers/NonlinearSolverParameters.hpp [hunk 7]
```diff
@@ -106,8 +107,18 @@ class NonlinearSolverParameters : public dataRepository::Group
...
 
+  /**
+   * @brief Indicates the handling of line search in a Newton loop.
+   */
+  enum class LineSearchAction : int
+  {
+    None,    ///< Do not use line search
+    Attempt, ///< Use line search. Allow exit from line search without achieving smaller residual than starting residual.
+    Require, ///< Use line search. If smaller residual than starting residual is not achieved, cut time step.
+  };
```
Comment (klevzoff) [inline src/coreComponents/physicsSolvers/NonlinearSolverParameters.hpp:118]: Do these sound reasonable?

**Comment by klevzoff** [comment]
@francoishamon rebased and ready to be checked with Intel

**Comment by klevzoff** [comment]
@francoishamon has found some `FAIL CHECK` tests on Quartz with intel, all coming from enums that did not explicitly specify their underlying type... and the compiler chosen type was obviously platform dependent, so they got serialized to different conduit types. We got away with this before, because it was always a 32-bit type (e.g. `int32` vs `uint32`) so if would always get serialized as `char[4]`. But if on some platform the compiler chose a 16-bit or a 64-bit type, we would have failed tests there even without this PR. 

TL;DR: the solution is to always specify the underlying type explicitly if the enum is to be registered with the repository. I've applied this to all enums currently registered, so should be good to go once another round of CI completes.
